earth_enterprise/src/common/geFilePoolImpl.h

```diff
@@ -20,18 +20,35 @@
 
 #include <khMTTypes.h>
 #include <fcntl.h>
+#include <string>
+#include <memory>
 
 class geFilePool;
 
+class AbstractFileIdentifier {
+public:
+  virtual bool isValid() { return false; }
+  virtual int getAsFD() { return -1; }
+  virtual void invalidate() {};
+};
+
+class POSIXFileIdentifier: public AbstractFileIdentifier {
+  int fd;
+public:
+  POSIXFileIdentifier(int i = -1) : fd{ i } {}
+  bool isValid() override { return fd != -1; }
+  void invalidate() override { fd = -1; }
+  int getAsFD() override { return fd; }
+};
 
 class FileReservationImpl : public khMTRefCounter {
   bool isWriter;
-  int fd;
+  std::shared_ptr<AbstractFileIdentifier> fid;
  public:
-  int Fd(void) const { return fd; }
-  FileReservationImpl(void) : isWriter(false), fd(-1) { }
+  std::shared_ptr<AbstractFileIdentifier> Fid(void) const { return fid; }
```

Comment:
You could probably use a `unique_ptr` here instead of a `shared_ptr` since `unique_ptr`s are cheaper. The different types of pointers represent different types of ownership (shared or unique/single, unsurprisingly). Here, ownership means who is supposed to delete the pointer.

- `shared_ptrs` are for when multiple objects own the same pointer. We use this in the cache - an item cannot be deleted until the cache and everything that has pulled the item out of the cache are done with it. Whoever has the last pointer is the one that deletes it.
- `unique_ptrs` are for when a single object owns the pointer and will eventually delete it. It looks like this example fits that case. Other objects might use the `fid` but this object owns it.

You can use `shared_ptrs` where `unique_ptrs` would work but it makes the code less efficient because it has to track how many references there are to the pointer.

In this case, I think you could change it to a `unique_ptr`. You'd also change the `Fid` function to return a raw pointer (`fid.get`) rather than the `unique_ptr`. Returning a `unique_ptr` means "I'm giving up ownership of this object and you now own it". Returning a raw pointer means "I still own this object but you can use it for a little while". It seems like the second case is what's happening here.

Actually, all of the `shared_ptrs` in this PR can possibly be changed to `unique_ptrs`, though that might require more refactoring.

There's more discussion about smart pointers and ownership here: https://www.modernescpp.com/index.php/c-core-guidelines-rules-to-smart-pointers.